prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt

```diff
@@ -1,135 +1,227 @@
+                error("For $userName, server2server scope $claimsScope: invalid format")
+            this.scopes = setOf(claimsScope)
+        }
+        this.isPrimeAdmin = scopes.contains(Scope.primeAdminScope)
+        // todo : have the TokenAuthentication.authenticate return type AuthenticatedClaims.
     }
 
     /**
-
-     * Use this for claims not coming from a human being (that is, not using Okta).
-     * For example, use this for claims for server-to-server auth.
-     * Each value passed must have been authenticated.  These are *not* for requested or required authorizations.
+     * Determine if these claims authorize access to the /validate and /waters endpoints for
+     * org [requiredSender] obj.  The [request] is only used for logging.
      *
-     * [_jwtClaims]  The raw list of claims.
-     * [_organizationNameClaim] The organization Name of this server. (Not the full name, eg oh-doh, not oh-doh.elr)
-     * [_isPrimeAdmin] true if this server claims to act as a PrimeAdmin
-     * [_isSenderOrgClaim] true if this claims set is from a Sender organization.  Certain queries (eg, give me
-     * a list of submissions) only make sense if the Organization has at least one Sender in it.
-     *
-     * The latter args are optional.   If not provided, this constructor will attempt to extract the values in the jwt
-     * claims, following our Okta claim forms (eg, "DH...", etc).   Otherwise it will default to least privilege.
+     * @return true if the request to submit data on behalf of [requiredSender]
+     * is authorized based on the [claims], false otherwise.
      */
-    constructor(
-        _jwtClaims: Map<String, Any>,
-        _organizationNameClaim: String? = null,
-        _isPrimeAdmin: Boolean? = null,
-        _isSenderOrgClaim: Boolean? = null,
-    ) {
-        this.jwtClaims = _jwtClaims
-        this.userName = _jwtClaims[oktaSubjectClaim]?.toString() ?: error("No username in claims")
-        @Suppress("UNCHECKED_CAST")
-        val memberships = jwtClaims[oktaMembershipClaim] as? Collection<String> ?: emptyList()
-        val orgNamePair = extractOrganizationNameFromOktaMembership(memberships)
-        this.isPrimeAdmin = _isPrimeAdmin ?: isPrimeAdmin(memberships)
-        this.organizationNameClaim = _organizationNameClaim ?: orgNamePair?.first // might be null
-        this.isSenderOrgClaim = _isSenderOrgClaim ?: orgNamePair?.second ?: false
+    fun authorizedForSubmission(
+        requiredSender: Sender,
+        request: HttpRequestMessage<String?>
+    ): Boolean {
+        return authorizedForSubmission(requiredSender.organizationName, requiredSender.name, request)
     }
 
     /**
-     * Derive useful info from jwtClaims on [memberships].
-     * @return the *first* well-formed ReportStream organizationName found in [memberships].
-     * (So this won't work if user has many organization claims.)
+     * Determine if these claims authorize access to the /validate and /waters endpoints for
```

Comment:
Update comment to include /report endpoint as well, since that is a submission endpoint

Reply:
Fixed